plugins/draco/io/DracoWriter.cpp

```diff
@@ -0,0 +1,457 @@
+    Dimension::Type pdalDataType = Dimension::defaultType(pt);
+    //get corresponding draco data type
+    draco::DataType dracoDataType = typeMap.at(pdalDataType);
+
+    //add generic type to the pointcloud
+    draco::GeometryAttribute ga;
+    ga.Init(draco::GeometryAttribute::GENERIC, nullptr, n, dracoDataType,
+            false, draco::DataTypeLength(dracoDataType) * n, 0);
+    auto attId = m_pc->AddAttribute(ga, true, m_pc->num_points());
+
+    //add generic attribute to map with its id
+    m_genericMap[pt] = attId;
+
+    //create metadata object and add dimension string as name
+    draco::Metadata metadata;
+    const std::string attName = "name";
+    const std::string name = Dimension::name(pt);
+    metadata.AddEntryString(attName, name);
+    std::unique_ptr<draco::AttributeMetadata> metaPtr(new draco::AttributeMetadata(metadata));
+
+    //attach metadata to generic attribute in pointcloud
+    m_pc->AddAttributeMetadata(attId, std::move(metaPtr));
+}
+
+void DracoWriter::addAttribute(draco::GeometryAttribute::Type t, int n) {
+    // - iterate over values in dimMap, which are draco dimensions
+    // - use the pdal type associated with it to get the pdal data type
+    // - use map of pdal types to draco types to get the correct draco type
+    // - add it to the point cloud and add attribute id to attMap with dim as key
+    const auto it = std::find_if(
+        dimMap.begin(),
+        dimMap.end(),
+        [t](const std::pair<Dimension::Id, draco::GeometryAttribute::Type>& one)
+        {
+            return one.second == t;
+        });
+
+    pdal::Dimension::Type pdalType = Dimension::defaultType(it->first);
+    draco::DataType dataType = draco::DT_INVALID;
+    if (it != dimMap.end())
+        dataType = typeMap.at(pdalType);
+
+    //create geometry attribute and add to pointcloud
+    draco::GeometryAttribute ga;
+    ga.Init(t, nullptr, n, dataType, false, draco::DataTypeLength(dataType) * n, 0);
+    m_attMap[t] = m_pc->AddAttribute(ga, true, m_pc->num_points());
+}
+
+void DracoWriter::initPointCloud(point_count_t size)
+{
+    //begin initialization of the point cloud with the point count
+    unsigned int sizeInt = size;
+    m_pc->set_num_points(sizeInt);
+    //go through known draco attributes and add to the pointcloud
+    for (auto &dim: m_dims)
+    {
+        addAttribute(dim.first, dim.second);
+    }
+    //do the same for generic attributes
+    for (auto &dim: m_genericDims)
+    {
+        addGeneric(dim, 1);
+    }
+}
+
+void DracoWriter::addPoint(int attId, Dimension::IdList idList, PointRef &point, PointId idx)
+{
+    //get first id in list. All ids in any given list should be the same datatype
+    Dimension::Id dim = idList[0];
+    Dimension::Type type;
+
+    if (m_userDimMap.find(dim) != m_userDimMap.end()) {
+        std::string typeStr = m_userDimMap.at(dim);
+        type = Dimension::type(typeStr);
+    } else {
+        type = Dimension::defaultType(dim);
+    }
+
```

Comment:
Just like in the reader, all this can be dispensed with by using the pdal function that will allow you to read data of the type you desire directly into a buffer.